src/scanners/javascript.js

```diff
@@ -272,22 +272,38 @@ export default class JavaScriptScanner {
       detected.sourceType = filename.endsWith('.mjs')
         ? 'module'
         : this._getSourceType(ast);
+
+      // If a .js file contains import.meta and wasn't previously detected as a
+      // module, parse it with espree to identify if import.meta causes an error
+      if (
+        this.code.includes('import.meta') &&
+        filename.endsWith('.js') &&
+        detected.sourceType === 'script'
+      ) {
+        espree.parse(this.code, { ...parserOptions, sourceType: 'script' });
+      }
```

Comment:
The first parse is always successful because the default sourceType is 'module'. It is necessary for sourceType to be 'module' `this._getSourceType(ast)` to run successfully.

Reply:
I see.  I think we overlooked a thing or two when we thought checking the error was the right thing to do.

I took a closer look:

- `_getSourceType()` does not help us here: if we parse successfully as a module (the default as mentioned above), we still get a "script" type because of `_getSourceType()`. This was the approach preferred in https://github.com/mozilla/addons-linter/issues/3639#issuecomment-806524060, mainly to minimize side effects
- The error reported in https://github.com/mozilla/addons-linter/issues/4331 is actually thrown by ESLint: https://github.com/mozilla/addons-linter/blob/d757265a8816d405e0531a78262d6cc694f4b0f7/src/scanners/javascript.js#L139-L151
- There is no parsing error in `detectSourceType()`, which is more or less where we wanted to check the error

I think it would be OK to add a condition like `this.code.includes('import.meta')` to force the source type to be `module`, e.g what this patch does (which also removes the parameter because it is useless):

```diff
diff --git a/src/scanners/javascript.js b/src/scanners/javascript.js
index d9550846..4918b483 100644
--- a/src/scanners/javascript.js
+++ b/src/scanners/javascript.js
@@ -47,7 +47,7 @@ export default class JavaScriptScanner {
     // as part of tests.
     _rules = undefined,
   } = {}) {
-    const detectedSourceType = this.detectSourceType(this.filename);
+    const detectedSourceType = this.detectSourceType();
     this.sourceType = detectedSourceType.sourceType;
 
     const rules = {};
@@ -245,19 +245,19 @@ export default class JavaScriptScanner {
   }
 
   /*
-    Analyze the source-code by naively parsing the source code manually and
-    checking for module syntax errors in order to determine the source type of
-    the file.
-
-    This function returns an object with the source type (`script` or `module`)
-    and a non-null parsing error object when parsing has failed with the default
-    source type. The parsing error object contains the `error` message and the
-    source `type`.
-  */
-  detectSourceType(filename) {
+   * Analyze the source-code by naively parsing the source code manually and
+   * checking for module syntax errors and/or some features in the source code
+   * in order to determine the source type of the file.
+   *
+   * This function returns an object with the source type (`script` or
+   * `module`) and a non-null parsing error object when parsing has failed with
+   * the default source type. The parsing error object contains the `error`
+   * message and the source `type`.
+   */
+  detectSourceType() {
     // Default options taken from eslint/lib/linter:parse
     const parserOptions = {
-      filePath: filename,
+      filePath: this.filename,
       sourceType: 'module',
       ecmaVersion: ECMA_VERSION,
     };
@@ -269,9 +269,10 @@ export default class JavaScriptScanner {
 
     try {
       const ast = espree.parse(this.code, parserOptions);
-      detected.sourceType = filename.endsWith('.mjs')
-        ? 'module'
-        : this._getSourceType(ast);
+      detected.sourceType =
+        this.filename.endsWith('.mjs') || this.code.includes('import.meta')
+          ? 'module'
+          : this._getSourceType(ast);
     } catch (exc) {
       const line = exc.lineNumber || '(unknown)';
       const column = exc.column || '(unknown)';
diff --git a/tests/unit/scanners/test.javascript.js b/tests/unit/scanners/test.javascript.js
index 961326f9..2fd9949a 100644
--- a/tests/unit/scanners/test.javascript.js
+++ b/tests/unit/scanners/test.javascript.js
@@ -494,7 +494,7 @@ describe('JavaScript Scanner', () => {
       `;
 
       const jsScanner = new JavaScriptScanner(code, 'code.js');
-      const detectedSourceType = jsScanner.detectSourceType('code.js');
+      const detectedSourceType = jsScanner.detectSourceType();
 
       expect(detectedSourceType.sourceType).toEqual('script');
       expect(detectedSourceType.parsingError).toEqual({
@@ -507,7 +507,7 @@ describe('JavaScript Scanner', () => {
       const code = `(function () {})();`;
       const jsScanner = new JavaScriptScanner(code, 'code.js');
 
-      const detectedSourceType = jsScanner.detectSourceType('code.js');
+      const detectedSourceType = jsScanner.detectSourceType();
 
       expect(detectedSourceType.sourceType).toEqual('script');
       expect(detectedSourceType.parsingError).toEqual(null);
@@ -518,7 +518,7 @@ describe('JavaScript Scanner', () => {
       const jsScanner = new JavaScriptScanner(code, 'code.js');
       sinon.stub(jsScanner, '_getSourceType').throws(new Error('some error'));
 
-      const detectedSourceType = jsScanner.detectSourceType('code.js');
+      const detectedSourceType = jsScanner.detectSourceType();
 
       expect(detectedSourceType.sourceType).toEqual('script');
       expect(detectedSourceType.parsingError).toEqual({
```


@rpl WDYT?

Reply:
As discussed in our triage meeting, I'll be also ok with this approach in the short term.

As a separate followup we are considering of reversing the order of the sourceType we use, going for "module" first and if parsing fails eventually fallback to "script".